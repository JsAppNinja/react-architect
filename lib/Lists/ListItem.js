'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactAddonsTransitionGroup = require('react-addons-transition-group');

var _reactAddonsTransitionGroup2 = _interopRequireDefault(_reactAddonsTransitionGroup);

var _reactAddonsPureRenderMixin = require('react-addons-pure-render-mixin');

var _reactAddonsPureRenderMixin2 = _interopRequireDefault(_reactAddonsPureRenderMixin);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Buttons = require('../Buttons');

var _Transitions = require('../Transitions');

var _Transitions2 = _interopRequireDefault(_Transitions);

var _List = require('./List');

var _List2 = _interopRequireDefault(_List);

var _ListItemText = require('./ListItemText');

var _ListItemText2 = _interopRequireDefault(_ListItemText);

var _Inks = require('../Inks');

var _Inks2 = _interopRequireDefault(_Inks);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ListItem = function (_Component) {
  _inherits(ListItem, _Component);

  function ListItem(props) {
    _classCallCheck(this, ListItem);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(ListItem).call(this, props));

    _this.renderLeftChildren = function () {
      var _this$props = _this.props;
      var leftIcon = _this$props.leftIcon;
      var leftAvatar = _this$props.leftAvatar;

      if (!leftIcon && !leftAvatar) {
        return null;
      }

      return _react2.default.cloneElement(leftIcon || leftAvatar, { key: 'left-children' });
    };

    _this.renderRightChildren = function () {
      var _this$props2 = _this.props;
      var rightIcon = _this$props2.rightIcon;
      var rightAvatar = _this$props2.rightAvatar;
      var expanderIconChildren = _this$props2.expanderIconChildren;
      var expanderIconClassName = _this$props2.expanderIconClassName;
      var nestedItems = _this$props2.nestedItems;
      var disabled = _this$props2.disabled;

      if (!rightIcon && !rightAvatar && !(nestedItems && nestedItems.length)) {
        return null;
      }

      if (nestedItems && nestedItems.length) {
        var className = (0, _classnames2.default)('md-list-expander', { 'active': _this.isOpen() });
        if (!rightIcon) {
          return _react2.default.createElement(_Buttons.IconButton, {
            key: 'toggle',
            disabled: disabled,
            onClick: _this.toggleNestedItems,
            iconClassName: expanderIconClassName,
            className: className,
            children: expanderIconChildren
          });
        }

        return _react2.default.cloneElement(rightIcon, { key: 'toggle', className: className });
      }

      return _react2.default.cloneElement(rightIcon || rightAvatar, { key: 'right-children' });
    };

    _this.isOpen = function () {
      return typeof _this.props.isOpen === 'undefined' ? _this.state.isOpen : _this.props.isOpen;
    };

    _this.toggleNestedItems = function (e) {
      var onExpanderClick = _this.props.onExpanderClick;

      e.stopPropagation();

      if (onExpanderClick) {
        onExpanderClick(e);
      } else {
        _this.setState({ isOpen: !_this.state.isOpen });
      }
    };

    _this.handleClick = function (e) {
      var _this$props3 = _this.props;
      var onClick = _this$props3.onClick;
      var nestedItems = _this$props3.nestedItems;
      var expandOnClick = _this$props3.expandOnClick;
      var disabled = _this$props3.disabled;

      if (disabled) {
        return;
      }
      onClick && onClick(e);

      if (expandOnClick && nestedItems) {
        _this.toggleNestedItems(e);
      }
    };

    _this.shouldComponentUpdate = _reactAddonsPureRenderMixin2.default.shouldComponentUpdate.bind(_this);
    _this.state = { isOpen: props.initiallyOpen, hover: false };
    return _this;
  }

  _createClass(ListItem, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var hover = this.state.hover;
      var _props = this.props;
      var component = _props.component;
      var className = _props.className;
      var containerClassName = _props.containerClassName;
      var primaryText = _props.primaryText;
      var secondaryText = _props.secondaryText;
      var threeLines = _props.threeLines;
      var leftIcon = _props.leftIcon;
      var leftAvatar = _props.leftAvatar;
      var rightIcon = _props.rightIcon;
      var rightAvatar = _props.rightAvatar;
      var nestedItems = _props.nestedItems;
      var expanderIconClassName = _props.expanderIconClassName;
      var expanderIconChildren = _props.expanderIconChildren;
      var disabled = _props.disabled;

      var props = _objectWithoutProperties(_props, ['component', 'className', 'containerClassName', 'primaryText', 'secondaryText', 'threeLines', 'leftIcon', 'leftAvatar', 'rightIcon', 'rightAvatar', 'nestedItems', 'expanderIconClassName', 'expanderIconChildren', 'disabled']);

      var text = _react2.default.createElement(_ListItemText2.default, {
        key: 'text',
        primaryText: primaryText,
        secondaryText: secondaryText,
        className: (0, _classnames2.default)({
          'avatar-offset': !!leftAvatar,
          'icon-offset': !!leftIcon
        })
      });

      var content = _react2.default.createElement(component, _extends({
        role: 'button',
        tabIndex: component === 'div' && !nestedItems ? 0 : null
      }, props, {
        disabled: disabled,
        onClick: this.handleClick,
        className: (0, _classnames2.default)('md-list-tile', className, {
          'secondary-action': nestedItems && nestedItems.length,
          'avatar-height': leftAvatar || rightAvatar,
          'two-lines': secondaryText,
          'three-lines': threeLines && secondaryText
        })
      }), [this.renderLeftChildren(), text, this.renderRightChildren()]);

      content = _react2.default.createElement(
        _Inks2.default,
        { disabled: disabled },
        content
      );

      var children = undefined;
      if (this.isOpen() && nestedItems && nestedItems.length) {
        children = _react2.default.createElement(
          _Transitions2.default,
          { key: 'nested-list' },
          _react2.default.createElement(
            _List2.default,
            null,
            nestedItems
          )
        );
      }

      return _react2.default.createElement(
        _reactAddonsTransitionGroup2.default,
        {
          component: 'li',
          className: (0, _classnames2.default)('md-list-item', containerClassName, { hover: hover }),
          onMouseOver: function onMouseOver() {
            return _this2.setState({ hover: true });
          },
          onMouseLeave: function onMouseLeave() {
            return _this2.setState({ hover: false });
          }
        },
        content,
        children
      );
    }
  }]);

  return ListItem;
}(_react.Component);

ListItem.propTypes = {
  primaryText: _react.PropTypes.node,
  secondaryText: _react.PropTypes.node,
  className: _react.PropTypes.string,
  containerClassName: _react.PropTypes.string,
  leftIcon: _react.PropTypes.node,
  leftAvatar: _react.PropTypes.node,
  rightIcon: _react.PropTypes.node,
  rightAvatar: _react.PropTypes.node,
  component: _react.PropTypes.oneOfType([_react.PropTypes.func, _react.PropTypes.string]).isRequired,
  nestedItems: _react.PropTypes.arrayOf(_react.PropTypes.node),
  initiallyOpen: _react.PropTypes.bool,
  isOpen: _react.PropTypes.bool,
  onExpanderClick: _react.PropTypes.func,
  expandOnClick: _react.PropTypes.bool,
  expanderIconChildren: _react.PropTypes.node,
  expanderIconClassName: _react.PropTypes.string,
  onClick: _react.PropTypes.func,
  disabled: _react.PropTypes.bool,
  threeLines: _react.PropTypes.bool
};
ListItem.defaultProps = {
  component: 'div',
  initiallyOpen: false,
  expanderIconChildren: 'keyboard_arrow_down',
  expandOnClick: true
};
exports.default = ListItem;
module.exports = exports['default'];