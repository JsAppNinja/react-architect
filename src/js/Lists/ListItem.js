import React, { Component, PropTypes } from 'react';
import PureRenderMixin from 'react-addons-pure-render-mixin';
import classnames from 'classnames';

import Ink from '../Inks';

export default class ListItem extends Component {
  constructor(props) {
    super(props);

    this.shouldComponentUpdate = PureRenderMixin.shouldComponentUpdate.bind(this);
    this.state = { focused: false };
  }

  static propTypes = {
    primaryText: PropTypes.node,
    secondaryText: PropTypes.node,
    secondaryText2: PropTypes.node,
    className: PropTypes.string,
    leftIcon: PropTypes.node,
    leftAvatar: PropTypes.node,
    rightIcon: PropTypes.node,
    rightAvatar: PropTypes.node,
    component: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.string,
    ]).isRequired,
  };

  static defaultProps = {
    component: 'li',
  };

  renderText = () => {
    const { primaryText, secondaryText, secondaryText2, leftIcon, leftAvatar, rightIcon, rightAvatar } = this.props;
    const tileTitle = <div key="tile-title" className="md-tile-primary-text">{primaryText}</div>;

    if(!leftIcon && !leftAvatar && !rightIcon && !rightAvatar) {
      return tileTitle;
    }

    const contentClassName = classnames('md-tile-content', {
      'icon-left': !!leftIcon || !!leftAvatar,
      'icon-right': !!rightIcon || !!rightAvatar,
    });
    return (
      <div key="tile-content" className={contentClassName}>
        {tileTitle}
        {secondaryText && <div className="md-tile-secondary-text">{secondaryText}</div>}
        {secondaryText2 && <div className="md-tile-secondary-text">{secondaryText2}</div>}
      </div>
    );
  };

  renderLeftChildren = () => {
    const { leftIcon, leftAvatar } = this.props;
    if(!leftIcon && !leftAvatar) {
      return null;
    }

    return React.cloneElement(leftIcon || leftAvatar, { key: 'left-children' });
  };

  renderRightChildren = () => {
    const { rightIcon, rightAvatar } = this.props;
    if(!rightIcon && !rightAvatar) { return null; }

    return React.cloneElement(rightIcon || rightAvatar, { key: 'right-children' });
  };

  render() {
    const { component, className, secondaryText, secondaryText2, leftIcon, leftAvatar, rightIcon, rightAvatar, ...props } = this.props;

    const item = React.createElement(component, {
      ...props,
      role: 'button',
      className: classnames('md-list-tile', className, {
        'md-list-3-lines': !!secondaryText && !!secondaryText2,
        'md-list-avatar': leftIcon || leftAvatar || rightIcon || rightAvatar,
      }),
    }, [this.renderLeftChildren(), this.renderText(), this.renderRightChildren()]);
    return (
      <Ink>
        {item}
      </Ink>
    );
  }
}
