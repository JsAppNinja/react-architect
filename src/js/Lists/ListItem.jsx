import React, { Component, PropTypes } from 'react';
import PureRenderMixin from 'react-addons-pure-render-mixin';
import classnames from 'classnames';

import { rippleComponent } from '../utils/Wrappers';

class ListItem extends Component {
  constructor(props) {
    super(props);

    this.shouldComponentUpdate = PureRenderMixin.shouldComponentUpdate.bind(this);
  }

  static propTypes = {
    primaryText: PropTypes.string.isRequired,
    secondaryText: PropTypes.node,
    secondaryText2: PropTypes.node,
    className: PropTypes.string,
    leftIcon: PropTypes.node,
    leftAvatar: PropTypes.node,
    rightIcon: PropTypes.node,
    rightAvatar: PropTypes.node,
    component: PropTypes.func,
  }

  renderText = () => {
    const { primaryText, secondaryText, secondaryText2, leftIcon, leftAvatar, rightIcon, rightAvatar } = this.props;
    const tileTitle = <div key="tile-title" className="md-tile-primary-text">{primaryText}</div>;

    if(!leftIcon && !leftAvatar && !rightIcon) {
      return tileTitle;
    }

    const contentClassName = classnames('md-tile-content', {
      'icon-left': !!leftIcon || !!leftAvatar,
      'icon-right': !!rightIcon || !!rightAvatar,
    });
    return (
      <div key="tile-content" className={contentClassName}>
        {tileTitle}
        {secondaryText && <div className="md-tile-secondary-text">{secondaryText}</div>}
        {secondaryText2 && <div className="md-tile-secondary-text">{secondaryText2}</div>}
      </div>
    );
  }

  renderLeftChildren = () => {
    const { leftIcon, leftAvatar } = this.props;
    if(!leftIcon && !leftAvatar) {
      return null;
    }

    return React.cloneElement(leftIcon || leftAvatar, { key: 'left-children' });
  }

  renderRightChildren = () => {
    return null;
  }

  render() {
    const { component, className, secondaryText, secondaryText2, ...props } = this.props;
    return React.createElement(component || 'li', {
      role: 'button',
      className: classnames('md-list-tile', className, { 'md-list-3-lines': !!secondaryText && !!secondaryText2 }),
      ...props,
    }, [this.renderLeftChildren(), this.renderText(), this.renderRightChildren()]);
  }
}
export default rippleComponent()(ListItem);
