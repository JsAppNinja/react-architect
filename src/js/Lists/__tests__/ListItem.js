/* eslint-env jest */
jest.unmock('../ListItem');

import React from 'react';
import { findDOMNode } from 'react-dom';
import { mount } from 'enzyme';
import {
  renderIntoDocument,
  findRenderedComponentWithType,
} from 'react-dom/test-utils';

import ListItem from '../ListItem';
import AccessibleFakeInkedButton from '../../Helpers/AccessibleFakeInkedButton';

describe('ListItem', () => {
  it('merges className and style', () => {
    const props = {
      style: { background: 'black' },
      className: 'test',
      tileStyle: { background: 'red' },
      tileClassName: 'womba-juice',
      primaryText: 'Woop',
    };

    const listItem = renderIntoDocument(<ListItem {...props} />);
    const listItemNode = findDOMNode(listItem);
    const btn = findRenderedComponentWithType(listItem, AccessibleFakeInkedButton);

    expect(listItemNode.style.background).toBe(props.style.background);
    expect(listItemNode.className).toContain(props.className);
    expect(btn.props.style).toEqual(props.tileStyle);
    expect(btn.props.className).toContain(props.tileClassName);
  });

  it('passes event listeners to the AccessibleFakeInkedButton', () => {
    const onMouseDown = jest.fn();
    const onMouseUp = jest.fn();
    const onDoubleClick = jest.fn();
    const onTouchMove = jest.fn();
    const onTouchCancel = jest.fn();
    const onFocus = jest.fn();
    const onBlur = jest.fn();

    const props = {
      primaryText: 'Test',
      onMouseDown,
      onMouseUp,
      onDoubleClick,
      onTouchMove,
      onTouchCancel,
      onFocus,
      onBlur,
    };

    const item = renderIntoDocument(<ListItem {...props} />);
    const btn = findRenderedComponentWithType(item, AccessibleFakeInkedButton);
    expect(btn.props.onMouseDown).toBe(props.onMouseDown);
    expect(btn.props.onMouseUp).toBe(props.onMouseUp);
    expect(btn.props.onDoubleClick).toBe(props.onDoubleClick);
    expect(btn.props.onTouchMove).toBe(props.onTouchMove);
    expect(btn.props.onTouchCancel).toBe(props.onTouchCancel);
    expect(btn.props.onFocus).toBe(props.onFocus);
    expect(btn.props.onBlur).toBe(props.onBlur);
  });

  it('calls the onClick prop when _handleClick is called', () => {
    const props = { primaryText: 'Test', onClick: jest.fn() };
    const item = renderIntoDocument(<ListItem {...props} />);
    const clickEvent = { target: {} };
    item._handleClick(clickEvent);

    expect(props.onClick.mock.calls.length).toBe(1);
    expect(props.onClick.mock.calls[0][0]).toEqual(clickEvent);
  });

  it('calls the onMouseOver prop when _handleMouseOver is called', () => {
    const props = { primaryText: 'Test', onMouseOver: jest.fn() };
    const item = renderIntoDocument(<ListItem {...props} />);
    const mouseEvent = { target: {} };
    item._handleMouseOver(mouseEvent);

    expect(props.onMouseOver.mock.calls.length).toBe(1);
    expect(props.onMouseOver.mock.calls[0][0]).toEqual(mouseEvent);
  });

  it('calls the onMouseLeave prop when _handleMouseLeave is called', () => {
    const props = { primaryText: 'Test', onMouseLeave: jest.fn() };
    const item = renderIntoDocument(<ListItem {...props} />);
    const mouseEvent = { target: {} };
    item._handleMouseLeave(mouseEvent);

    expect(props.onMouseLeave.mock.calls.length).toBe(1);
    expect(props.onMouseLeave.mock.calls[0][0]).toEqual(mouseEvent);
  });

  it('calls the onTouchStart prop when _handleTouchStart is called', () => {
    const props = { primaryText: 'Test', onTouchStart: jest.fn() };
    const item = renderIntoDocument(<ListItem {...props} />);
    const touchEvent = { target: {} };
    item._handleTouchStart(touchEvent);

    expect(props.onTouchStart.mock.calls.length).toBe(1);
    expect(props.onTouchStart.mock.calls[0][0]).toEqual(touchEvent);
  });

  it('calls the onTouchEnd prop when _handleTouchEnd is called', () => {
    const props = { primaryText: 'Test', onTouchEnd: jest.fn() };
    const item = renderIntoDocument(<ListItem {...props} />);
    const touchEvent = { target: {} };
    item._handleTouchEnd(touchEvent);

    expect(props.onTouchEnd.mock.calls.length).toBe(1);
    expect(props.onTouchEnd.mock.calls[0][0]).toEqual(touchEvent);
  });

  it('calls the onKeyDown prop when _handleKeyDown is called', () => {
    const props = { primaryText: 'Test', onKeyDown: jest.fn() };
    const item = renderIntoDocument(<ListItem {...props} />);
    const keyEvent = { target: {} };
    item._handleKeyDown(keyEvent);

    expect(props.onKeyDown.mock.calls.length).toBe(1);
    expect(props.onKeyDown.mock.calls[0][0]).toEqual(keyEvent);
  });

  it('calls the onKeyUp prop when _handleKeyUp is called', () => {
    const props = { primaryText: 'Test', onKeyUp: jest.fn() };
    const item = renderIntoDocument(<ListItem {...props} />);
    const keyEvent = { target: {} };
    item._handleKeyUp(keyEvent);

    expect(props.onKeyUp.mock.calls.length).toBe(1);
    expect(props.onKeyUp.mock.calls[0][0]).toEqual(keyEvent);
  });

  it('should correctly apply the itemProps and tileProps props', () => {
    const console = global.console;
    global.console = {
      ...console,
      error: (e) => {
        // don't care about unknown props for the newly added one
        if (!e.match(/Unknown prop/)) {
          console.error(e);
        }
      },
    };
    const otherProp = 'other-prop';
    const itemProps = { onClick: jest.fn() };
    const tileProps = { onMouseDown: jest.fn() };
    const item = mount(
      <ListItem
        primaryText="Test"
        itemProps={itemProps}
        tileProps={tileProps}
        otherProp={otherProp}
        passPropsToItem={false}
      />
    );
    let li = item.find('li');
    let button = item.find(AccessibleFakeInkedButton);
    expect(li.props().onClick).toBe(itemProps.onClick);
    expect(button.props().onMouseDown).toBe(tileProps.onMouseDown);
    expect(button.props().otherProp).toBe('other-prop');

    item.setProps({ passPropsToItem: true });
    li = item.find('li');
    button = item.find(AccessibleFakeInkedButton);
    expect(li.props().onClick).toBe(itemProps.onClick);
    expect(li.props().otherProp).toBe('other-prop');
    expect(button.props().onMouseDown).toBe(tileProps.onMouseDown);

    global.console = console;
  });
});
