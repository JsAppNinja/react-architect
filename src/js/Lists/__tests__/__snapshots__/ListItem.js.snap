// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem should render correctly with a snapshot tester 1`] = `
<li
  aria-posinset={undefined}
  aria-setsize={undefined}
  className="md-list-item"
  style={undefined}
>
  <div
    aria-expanded={null}
    aria-pressed={false}
    className="md-fake-btn md-pointer--hover md-fake-btn--no-outline md-list-tile md-text"
    disabled={undefined}
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    role="button"
    style={undefined}
    tabIndex={0}
  >
    <div
      className="md-ink-container"
      style={undefined}
    />
    <div
      className="md-tile-content"
      style={undefined}
    >
      <div
        className="md-tile-text--primary md-text"
        style={undefined}
      >
        Test
      </div>
    </div>
  </div>
</li>
`;
